ludwig/utils/dataframe_utils.py

```diff
@@ -108,3 +108,28 @@ def to_batches(df: pd.DataFrame, batch_size: int) -> List[pd.DataFrame]:
 @DeveloperAPI
 def from_batches(batches: List[pd.DataFrame]) -> pd.DataFrame:
     return pd.concat(batches)
+
+
+@DeveloperAPI
+def to_scalar_df(df: pd.DataFrame) -> pd.DataFrame:
+    """Converts all columns in a pd.DataFrame to be scalar types.
+
+    For object columns of lists, each element of the list is expanded into its own column named {column}_{index}. We
+    assume all object columns are lists of the same length (i.e., tensor format output from preprocessing). It's also
+    important that the relative order of the columns is preserved, to maintain consistency with other conversions like
+    the one for Hummingbird.
+    """
+    scalar_df = df
+    column_ordering = []
```

Comment:
For my own curiosity: Do we need to maintain the preexisting column order? If not, we could process only the object columns by with `df.select_dtype(include="object")`, drop the object columns from the original df, and concat the old and new scalar columns.

Reply:
The column order needs to be consistent with the way the conversion is done for hummingbird, which uses this ordering. We could change it for hummingbird, but probably 6 to one, half dozen to the other.

Reply:
Fair enough!